toolkit/offchain/src/d_param/mod.rs

```diff
@@ -0,0 +1,316 @@
+	types::OgmiosUtxo,
+};
+use partner_chains_plutus_data::d_param::{d_parameter_to_plutus_data, DParamDatum};
+use sidechain_domain::{DParameter, McTxHash, UtxoId};
+
+#[cfg(test)]
+mod tests;
+
+pub async fn upsert_d_param<C: QueryLedgerState + QueryNetwork + Transactions>(
+	genesis_utxo: UtxoId,
+	d_parameter: &DParameter,
+	payment_signing_key: [u8; 32],
+	ogmios_client: &C,
+) -> anyhow::Result<Option<McTxHash>> {
+	let ctx = TransactionContext::for_payment_key(payment_signing_key, ogmios_client).await?;
+	let (validator, policy) = crate::scripts_data::d_parameter_scripts(genesis_utxo, ctx.network)?;
+	let validator_address = validator.address_bech32(ctx.network)?;
+	let validator_utxos = ogmios_client.query_utxos(&[validator_address]).await?;
+
+	match get_current_d_parameter(validator_utxos)? {
+		Some((_, current_d_param)) if current_d_param == *d_parameter => {
+			log::info!("Current D-parameter value is equal to the one to be set.");
+			Ok(None)
+		},
+		Some((current_utxo, _)) => {
+			log::info!("Current D-parameter is different to the one to be set. Updating.");
+			Ok(Some(
+				update_d_param(
+					&validator,
+					&policy,
+					d_parameter,
+					&current_utxo,
+					ctx,
+					genesis_utxo,
+					ogmios_client,
+				)
+				.await?,
+			))
+		},
+		None => {
+			log::info!("There is no D-parameter set. Inserting new one.");
+			Ok(Some(
+				insert_d_param(&validator, &policy, d_parameter, ctx, genesis_utxo, ogmios_client)
+					.await?,
+			))
+		},
+	}
+}
+
+fn get_current_d_parameter(
+	validator_utxos: Vec<OgmiosUtxo>,
+) -> Result<Option<(OgmiosUtxo, DParameter)>, anyhow::Error> {
+	if let Some(utxo) = validator_utxos.first() {
+		let datum = utxo.datum.clone().ok_or_else(|| {
+			anyhow!("Invalid state: an UTXO at the validator script address does not have a datum")
+		})?;
+		let datum_plutus_data = PlutusData::from_bytes(datum.bytes).map_err(|e| {
+			anyhow!("Internal error: could not decode datum of D-parameter validator script: {}", e)
+		})?;
+		let current_d_param: DParameter =
+			DParamDatum::try_from(datum_plutus_data)
+				.map_err(|e| {
+					anyhow!("Internal error: could not decode datum of D-parameter validator script: {}", e)
+				})?
+				.into();
+		Ok(Some((utxo.clone(), current_d_param)))
+	} else {
+		Ok(None)
+	}
+}
+
+async fn insert_d_param<C: QueryLedgerState + Transactions + QueryNetwork>(
+	validator: &PlutusScript,
+	policy: &PlutusScript,
+	d_parameter: &DParameter,
+	ctx: TransactionContext,
+	genesis_utxo: UtxoId,
+	client: &C,
+) -> anyhow::Result<McTxHash>
+where
+	C: Transactions,
+{
+	let zero_ex_units = ScriptExUnits {
+		mint_ex_units: vec![
+			ExUnits::new(&0u64.into(), &0u64.into()),
+			ExUnits::new(&0u64.into(), &0u64.into()),
+		],
+		spend_ex_units: vec![],
+	};
+
+	let gov_utxo = crate::init_governance::get_governance_utxo(genesis_utxo, client)
+		.await
+		.map_err(|e| JsError::from_str(e.to_string().as_str()))?;
```

Comment:
Very late comment and no change required in this PR, but I think that `get_goveranance_utxo` should return an `Option`. Just wanting to hear your opinion.

Reply:
Anyway, the change here should be different message, that would point to the problem - `init-governance` wasn't executed.